### PR TITLE
Fix 94

### LINKED ISSUE #94: Can't refer to a type qualified with a module name
Given

    module m
    type t = int

The following yields a syntax error but should be allowed:

    module n
    type u = m.t

This is especially important since the changes discussed in #62 won't permit `m.t` to be imported but should still allow it to be referenced.

### PR DESCRIPTION
**Pull request by j14159**
Fixes #94 allowing us to qualify type constructors with their module name rather than always requiring a type to be imported.  This contributes to #62 by only allowing the use of constructors if their parent type has been exported by the module it's defined in.

### CODE REVIEW

File: src/alpaca_typer.erl [hunk 1]
```diff
@@ -1009,9 +1013,49 @@ inst_type_arrow(EnvIn, {type_constructor, Line, Name}) ->
...
+    %% If the constructor was not qualified with a module name it's pretty
+    %% to fetch but if it was, then we need to first make sure it's in
+    %% the specified module *and* that its enclosing type is exported.
+    %%
+    %% Here's the easy local get:
+    GetTC = fun(#type_constructor{line=Line, name=Name, module=undefined}) ->
+                    Default = {error, {bad_constructor, Line, Name}},
+                    %% constructors defined in this module or imported by it:
+                    Available = EnvIn#env.type_constructors,
+                    C = proplists:get_value(Name, Available, Default);
```
Comment (danabr) [inline src/alpaca_typer.erl]: Unused variable C (but maybe you added it for documentation purposes)?

File: src/alpaca_typer.erl [hunk 2]
```diff
@@ -1009,9 +1013,49 @@ inst_type_arrow(EnvIn, {type_constructor, Line, Name}) ->
...
+                            %% similar manner.
+                            Types = Target#alpaca_module.types,
+                            F = fun(#alpaca_type{members=Ms}=AT) ->
+                                        [M#alpaca_constructor{type=AT} ||
+                                            #alpaca_constructor{
+                                               name=#type_constructor{name=TCN}
+                                              }=M <- Ms, TCN =:= Name]
+                                     end,
+                            case lists:flatten(lists:map(F, Types)) of
+                                [RealC|_] ->
```
Comment (danabr) [inline src/alpaca_typer.erl]: Curious: When would you have multiple constructors with the same name?

Comment (j14159) [inline src/alpaca_typer.erl]: Good point, the AST gen state does ensure unique constructors within a module.  I was being overly cautious I think :)

File: src/alpaca_typer.erl (hunk 1 above)
Comment (j14159) [inline src/alpaca_typer.erl]: No, definite mistake, thanks for catching that :)

File: src/alpaca_typer.erl [hunk 3]
```diff
@@ -143,7 +143,7 @@ copy_cell(Cell, RefMap) ->
             {RevMembers, Map2} = lists:foldl(F, {[], RefMap}, Members),
             {RV2, Map3} = copy_cell(RV, Map2),
             {new_cell(#t_record{members=lists:reverse(RevMembers), row_var=RV2}), Map3};
-        #adt{vars=TypeVars, members=Members}=ADT ->
+        #adt{vars=TypeVars, members=Members, module=Mod}=ADT ->
```
Comment (j14159) [inline src/alpaca_typer.erl]: I think there are a couple of warnings coming from leex and yecc that might be tricky to squash but I could be wrong.  We could certainly try it.

Regardless, thanks for pointing this out.

File: src/alpaca_typer.erl [hunk 4]
```diff
@@ -986,13 +990,13 @@ inst_type_members(ADT, [_|T], Env, Memo) ->
...
                              {error, {bad_constructor_arg,term()}}.
-inst_type_arrow(EnvIn, {type_constructor, Line, Name}) ->
+inst_type_arrow(EnvIn, #type_constructor{}=TC) ->
     %% 20160603:  I have an awful lot of case ... of all over this
     %% codebase, trying a lineup of functions specific to this
     %% task here instead.  Sort of want Scala's `Try`.
     ADT_f = fun({error, _}=Err) ->
                     Err;
-               (#alpaca_constructor{type=#alpaca_type{}=T}=C) ->
+               (#alpaca_constructor{type=#alpaca_type{module=Mod}=T}=C) ->
```
Comment (j14159) [inline src/alpaca_typer.erl]: Removing, thanks!

**Pull request closed without merging**